pkg/jobs/registry.go

```diff
@@ -895,6 +899,141 @@ func (r *Registry) Start(ctx context.Context, stopper *stop.Stopper) error {
+	if r.knobs.DisableJobsMetricsPolling {
+		return nil
+	}
+
+	for {
+		const TxnName = "Jobs Stats Poller Txn"
+		txn := r.db.NewTxn(ctx, TxnName)
+
+		exists, err := JobExists(ctx, r.ex, txn, func(payload *jobspb.Payload) bool {
+			return payload.Type() == jobspb.TypePollJobsStats
+		})
+		if err != nil {
+			if errors.HasType(err, (*roachpb.TransactionRetryWithProtoRefreshError)(nil)) {
+				continue
+			}
+			return err
+		}
+
+		if exists {
+			return nil
+		}
+
+		metricsJobID := r.MakeJobID()
+		jr := createJobsStatsPollerRecord()
+		if _, err := r.CreateAdoptableJobWithTxn(ctx, jr, metricsJobID, txn); err != nil {
+			if errors.HasType(err, (*roachpb.TransactionRetryWithProtoRefreshError)(nil)) {
+				continue
+			}
+			return err
+		}
+		if err := txn.Commit(ctx); err != nil {
+			if errors.HasType(err, (*roachpb.TransactionRetryWithProtoRefreshError)(nil)) {
+				continue
+			}
+			return err
+		}
+		return nil
+	}
+}
+
+func createJobsStatsPollerRecord() Record {
+	return Record{
+		Description: jobspb.TypePollJobsStats.String(),
+		Details:     jobspb.PollJobsStatsDetails{},
+		Progress:    jobspb.PollJobsStatsProgress{},
+		CreatedBy:   &CreatedByInfo{Name: username.RootUser, ID: username.RootUserID},
+		Username:    username.RootUserName(),
+	}
+}
+
+// PollMetricsTask polls the jobs table for certain metrics at an interval.
+func (r *Registry) PollMetricsTask(ctx context.Context) error {
+
+	var err error
+	updateMetrics := func(ctx2 context.Context, s sqlliveness.Session) {
+		lc, cleanup := makeLoopController(r.settings, pollJobsMetricsInterval, r.knobs.IntervalOverrides.PollMetrics)
+		defer cleanup()
+		for {
+			select {
+			case <-ctx2.Done():
+				err = ctx.Err()
+				return
+			case <-lc.updated:
+				lc.onUpdate()
+			case <-lc.timer.C:
+				lc.timer.Read = true
+				if err = r.updatePausedMetrics(ctx2, s); err != nil {
+					log.Errorf(ctx2, "failed to update paused metrics: %v", err)
+					return
+				}
+				lc.onExecute()
+			}
+		}
+	}
+	r.withSession(ctx, updateMetrics)
+	return err
+}
+
+const pausedJobsCountQuery = string(`
+	SELECT job_type, count(*)
+	FROM crdb_internal.jobs
+	WHERE status = '` + StatusPaused + `' 
+  GROUP BY job_type`)
+
+func (r *Registry) updatePausedMetrics(ctx context.Context, s sqlliveness.Session) error {
+	metricUpdates := make(map[jobspb.Type]int)
+	err := r.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		// In case of transaction retries, reset this map here.
+		metricUpdates = make(map[jobspb.Type]int)
+
+		// Run the claim transaction at low priority to ensure that it does not
+		// contend with foreground reads.
+		if err := txn.SetUserPriority(roachpb.MinUserPriority); err != nil {
+			return errors.WithAssertionFailure(err)
+		}
+		// Note that we have to buffer all rows first - before processing each
+		// job - because we have to make sure that the query executes without an
+		// error (otherwise, the system.jobs table might diverge from the jobs
+		// registry).
```

Comment:
is the comment appropriate for this code or was it a copy/pasta?

Reply:
Copy pasta. The other could would modify the registry state during execution and not rollback those state changes if errors occurred while reading the jobs table. My code doesn't do that, so I took out the comment.